### PR TITLE
Fix session filters is None from request session

### PR DESCRIPTION
**Pull request by ChandlerBent**
If get session filters is `None` from request session, the search form will be raise no attribute 'cleaned_data' exception.

### CODE REVIEW

File: misago/admin/views/generic/list.py [hunk 1]
```diff
@@ -200,7 +200,8 @@ def get_filters_from_GET(self, request, search_form):
         return self.clean_filtering_data(form.cleaned_data)
 
     def get_filters_from_session(self, request, search_form):
-        session_filters = request.session.get(self.filters_session_key, {})
+        session_filters = request.session.get(
+            self.filters_session_key, {}) or {}
```
Comment (rafalp) [inline misago/admin/views/generic/list.py]:
```suggestion
             self.filters_session_key) or {}
```

**Comment by rafalp** [comment]
Can you provide test demonstrating the issue and fixing the crash? Or instruction how this crash can be reproduced from admin UI?

**Comment by ChandlerBent** [comment]
I set any field in User Accounts menu of search list, then I clear the search list. This is OK.
But I set again that I will get this exception.

**Comment by rafalp** [comment]
Can you add test to `misago/admin/tests/test_generic_admin_list_view.py` that does what you've described above and confirms no error is raised?

**Comment by ChandlerBent** [comment]
Yes, I will add the test later.

**Comment by ChandlerBent** [comment]
Done.

File: misago/admin/views/generic/list.py [hunk 2]
```diff
@@ -134,7 +134,7 @@ def dispatch(
             if request.GET.get("set_filters"):
                 # Force store filters in session
                 session_key = self.filters_session_key
-                request.session[session_key] = context["active_filters"]
+                request.session[session_key] = context["active_filters"] or {}
```
Comment (rafalp) [inline misago/admin/views/generic/list.py]: Why is this line changed too? Looks like bad place to fix `context` because few lines down we have other item access on `context` that will possibly fail if context is not a dict outside of this condition.

File: misago/admin/tests/test_generic_admin_list_view.py [hunk 3]
```diff
@@ -37,3 +37,12 @@ def test_unicode_is_preserved_in_redirect_querystring(admin_client):
 def test_view_is_not_redirecting_if_flag_is_set_in_querystring(admin_client):
     response = admin_client.get("%s?redirected=1" % list_link)
     assert response.status_code == 200
+
+
+def test_view_session_filter_is_none(admin_client):
```
Comment (rafalp) [inline misago/admin/tests/test_generic_admin_list_view.py]: Please make test name more descriptive:

- what state is tested?
- what is expected result?

Look at other tests names in module for inspiration, something like `test_restoring_filters_from_session_handles_filters_entry_being_none`

On sidenote, I am trying to find place where session filters are changed to `None`, but I can't seem to find one 🤔

File: misago/admin/tests/test_generic_admin_list_view.py [hunk 4]
```diff
@@ -37,3 +37,12 @@ def test_unicode_is_preserved_in_redirect_querystring(admin_client):
...
+    response = admin_client.get("%s?set_filters=1&rank=1" % list_link)
+    assert response.status_code == 302
+    response = admin_client.get("%s?set_filters=0" % list_link)
+    assert response.status_code == 302
+    response = admin_client.get("%s" % list_link)
```
Comment (rafalp) [inline misago/admin/tests/test_generic_admin_list_view.py]:
```suggestion
    response = admin_client.get("%s?redirected=1" % list_link)
```

File: misago/admin/views/generic/list.py (hunk 2 above)
Comment (ChandlerBent) [reply misago/admin/views/generic/list.py]: You're right.

File: misago/admin/tests/test_generic_admin_list_view.py (hunk 3 above)
Comment (ChandlerBent) [reply misago/admin/tests/test_generic_admin_list_view.py]: > Look at other tests names in module for inspiration, something like test_restoring_filters_from_session_handles_filters_entry_being_none

OK.

> On sidenote, I am trying to find place where session filters are changed to None, but I can't seem to find one 🤔
```python
            if request.GET.get("set_filters"):
                # Force store filters in session
                session_key = self.filters_session_key
                request.session[session_key] = context["active_filters"] or {}
                refresh_querystring = True
```
You can see my test. My second request has set_filters query params. Whether `set_filters` is `0` or `1`, it's still `str` type. So, the function will be into this part code. 

`context["active_filters"]` is `None` now. The problem is here.

**Comment by ChandlerBent** [comment]
I try my best to fix this bug.
But make more trouble.
Sorry to you.

My english not good.
The test function name so bad. 
:(

**Comment by rafalp** [comment]
You are doing great, don't worry! 👍

I am just wondering if you know which line in `list.py` changes `session[filters]` to `None`?

**Comment by ChandlerBent** [comment]
[Here](https://github.com/rafalp/Misago/pull/1177#discussion_r245466742)

**Comment by rafalp** [comment]
I saw this line, but I am trying to wrap my mind around it. Where `context["active_filters"]` becomes `Null` instead of empty `{}`?

**Comment by ChandlerBent** [comment]
```python
    def apply_filtering_on_context(self, context, active_filters, search_form):
        context["active_filters"] = active_filters
        context["search_form"] = search_form(initial=context["active_filters"])

        if context["active_filters"]:
            context["items"] = context["search_form"].filter_queryset(
                active_filters, context["items"]
            )
```
220 line

The `get_filtering_method_to_use` function is key.
```python
    def get_filtering_method_to_use(self, methods):
        for method in ("GET", "session"):
            if methods.get(method):
                return methods.get(method)
```
The `methods` context is `{'GET': {}, 'session': {}}`. The for part code either get `GET` or `session` is `{}`. So the function return None. It will be influence `context["active_filters"]` value.

Maybe `if methods.get(method):` should be correct as `if methods.get(method) is not None:`.
I don't know that is the correct good idea.

**Comment by rafalp** [comment]
Hello @ChandlerBent 

I've cloned your PR locally, investigated it and found where the bug lies ;)

1. Misago admin calls `get_filtering_method_to_use()`expecting it to return `dict` with filters that should be used to filter.
2. When querystring contains no valid filters, both `GET` and `session` filtering methods will return no filters, and `get_filtering_method_to_use()` will return `None`.
3. Because url has `set_filters`, this `None` will then be assigned as session filter.
4. On next request this `None` will be read from session, failing for and crashing admin.

Actual fix is adding the `return {}` at the end of `get_filtering_method_to_use()`:

```
def get_filtering_method_to_use(self, methods):
    for method in ("GET", "session"):
        if methods.get(method):
            return methods.get(method)
    return {}
```

**Comment by rafalp** [comment]
Thank you for making Misago better!

Your commits will land as #1190, I'll also backport the fix to Misago 0.19 today or tomorrow!

**Comment by rafalp** [comment]
Heads up @ChandlerBent I've noticed this bug was introduced on master branch when I was fixing `pylint` compliance errors in Misago: https://github.com/rafalp/Misago/commit/67838a3e8e12b333bf37db969890e2f44485e3aa#diff-97d5f581d1c4a4c697b0801d4065bab6L217

If you are running Misago master on production, please don't - we don't support that.
